Refactor code in adr_repo to prepare for future PRs

rustysys-dev opened the pull request:
Since the `list_all_adr` function sorts the list of adrs returned to `get_last_seq_id` the highest valued name should always be `adrs_paths[0]` so we can skip iterating through the entire list and just call `get_seq_id_from_name` on `adrs_paths[0]`.

Since `get_last_seq_id` was the only caller of `get_last_seq_id_from_all` we can remove the function.

If there was a reason I missed for `get_last_seq_id_from_all` please let me know.

rustysys-dev commented:
@omallassi  I am working to push another commit to fix the tests affected by this when I noticed that your tests are expecting the ability for adr src directories where adrs can be nested inside subdirectories?  However the tool doesn't seem to support this.  I opened up a new way to support this without having to perform any iterations over the adrs.  Should be something like:

- rename struct `Adr`s `file_name` property to `file_path`
- replace all references to `file_name` with `file_path`
- add logic to get the new `file_name` property from the `file_path` in the `Adr::from()` method. 
- add a new `file_name` property to use for sorting against in the `list_all_adr` function.

let me know if you have any thoughts

omallassi commented:
Regarding “ directories where adrs can be nested inside subdirectories?”

You are right . The read is “recursive” but the creation is not...
So Yes we should homogeneize  That is a great proposal.

rustysys-dev commented:
Summary: 

- Limited the scope of use cases required to be supported by `fn get_seq_id_by_name()` by adding a `file_name` parameter to `Adr`
- Added `file_id` parameter to `Adr`
- Added `file_path` parameter to `Adr` (was old `file_name`)
- Created a convenience function for sorting the result of `list_all_adr` by file_id`
- Created `Adr::new()` for creating default `Adr` object with sane defaults
- Updated `Adr` methods to be self mutating
- Fixed `Adr::from()` to support new properties
- Fixed any tests requiring new properties.
- Moved some commonly used test strings (templates) to test module constants
- Optimized `transition_to()` 
- Optimized `get_last_seq_id()`
- Deleted `get_seq_id_by_all()`

rustysys-dev commented:
@omallassi when you have time, if you could take a look.  (sorry its a lot :sweat_smile: ) 
This all should lay the groundwork for some of the other features I would like to add.

omallassi commented:
@rustysys-dev  - LGTM & thank you very much for this. 
the only "thing" is the cucumber tests are failing (at least on my side). From what I can see, this could be related to transition_to. 

what did you get when you run `cargo test` on your side? will try to have a look on my side. 

btw I am trying to plug statuses on github PR...

rustysys-dev commented:
I will look into the cucumber tests add a commit to fix them, I didn't take them into account.

rustysys-dev commented on adr_core_local_impl/src/adr_repo/mod.rs:
```diff
@@ -256,91 +239,78 @@ pub fn split_path<'a>(base_path: &'a Path, file_path: &'a Path) -> (&'a Path, &'
-                    //if status has been accepted
-                    match updated_adr_by_tuple.1 {
-                        true => {
-                            let updated_adr_from = updated_adr_from_tuple.0.add_reference(format!("{}", updated_adr_by_tuple.0.file_name).as_str());
-                            let updated_adr_by = updated_adr_by_tuple.0.add_reference(format!("{}", updated_adr_from.file_name).as_str());
```
found the issue, I forgot to add these `add_reference()` calls back in... :disappointed_relieved:

rustysys-dev replied:
making a commit to fix this now.

rustysys-dev commented on adr_core_local_impl/src/adr_repo/mod.rs:
```diff
@@ -256,91 +239,78 @@ pub fn split_path<'a>(base_path: &'a Path, file_path: &'a Path) -> (&'a Path, &'
-                            let updated_adr_from = updated_adr_from_tuple.0.add_reference(format!("{}", updated_adr_by_tuple.0.file_name).as_str());
-                            let updated_adr_by = updated_adr_by_tuple.0.add_reference(format!("{}", updated_adr_from.file_name).as_str());
-                
-                            fs::write(from, updated_adr_from.content)?;
```
fixed !

rustysys-dev commented:
@omallassi 

The PR seems to have passed the Github tests and 8/9 of the Travis CI tests.  

The Travis CI tests had adr_config test failure, something I noticed happens from time to time.  Is this a pre-existing issue? or is it possible that my changes affected the adr_config tests?
